src/mpid/ch4/netmod/ofi/ofi_win.c

```diff
@@ -168,11 +168,15 @@ static int win_allgather(MPIR_Win * win, void *base, int disp_unit)
      * we skip trial and return immediately. When FI_MR_ALLOCATED is not set, however,
      * it lacks documentation defining whether the provider can accept NULL and whether
      * it means registration of the entire virtual address space, although we have observed
-     * successful use in existing providers. Thus, we take a try here. */
-
+     * successful use in most of the existing providers. Thus, we take a try here for most
+     * providers. The CXI provider requires a non-zero virtual address and does not accept
+     * NULL base address. For this reason, when the base address is NULL with the CXI provider,
+     * we skip trial and return.*/
     int rc = 0, allrc = 0;
     MPIDI_OFI_WIN(win).mr = NULL;
-    if (base || (win->create_flavor == MPI_WIN_FLAVOR_DYNAMIC && !MPIDI_OFI_ENABLE_MR_ALLOCATED)) {
+    if (base || (win->create_flavor == MPI_WIN_FLAVOR_DYNAMIC &&
+                 !MPIDI_OFI_ENABLE_MR_ALLOCATED &&
+                 strcmp("cxi", MPIDI_OFI_global.prov_use[0]->fabric_attr->prov_name) != 0)) {
```

Comment:
the whole branch here is a trial, it is okay for `fi_mr_reg` to fail. The `rc` will be allreduced later and dealt with.

Reply:
The issue is with `MPIDI_OFI_mr_bind` on line 200, right? That call should be moved to the 'else-` branch (line 216).

Reply:
MPIDI_OFI_mr_bind does cause the failure. However, interestingly, with CXI, when a NULL address is passed info `fi_mr_reg`, rc is returned as 0. So, it is not treated as a failure to register. Maybe this is a CXI bug?

Reply:
I see. How about return rc from MPIDI_OFI_mr_bind instead?

Reply:
Currently, MPIDI_OFI_mr_bind returns `mpi_errno`, which is I think greater than zero. `fi_mr_enable` does return -22 when base is NULL, so I could transform the function to return that and use that in the allreduce. Does that sound good?

Reply:
Also, in this case if rc < 0, I see that it just closes MR and exits without calling `load_acc_hint`. And for collective attach programs, this causes a SIGSEGV when accessing `MPIDI_OFI_WIN(win).acc_hint`, so somehow that should be called too. This makes me think that the existing solution of using the else path on line 204 is not a bad idea.

Reply:
@hzhou Additionally, with this approach when I just tried it again with `mpiexec -n 4   -ppn 1 ./window_attach_collattach`, I saw a node failure. So, it looks like CXI definitely doesn't like NULL registration followed by bind.

Reply:
I see. Rather than do `strcmp` of the provider name during run time, we should do that check during init time and set a capability variable, e.g. `MPIDI_OFI_ENABLE_MR_REGISTER_NULL`, i.e. add a new capability variable rather than ad-hoc patches.

Reply:
Okay, that's a good solution. So, basically, I need to add this new macro in `ofi_capability_sets.h` and set it OFF for CXI since it doesn't support NULL MR registration and ON for all other providers, right?

Reply:
That sounds about right

Reply:
Also, I see from the libfabric wiki that FI_MR_ALLOCATED refers to: 

```
FI_MR_ALLOCATED
When set, all registered memory regions must be backed by physical memory pages at the time the registration call is made.
```

This makes me think that FI_MR_ALLOCATED should be set for CXI since it doesn't support registering NULL and only support registering actual physical memory regions. I am checking with the CXI engineers to find out why FI_MR_ALLOCATED is not set. 

If FI_MR_ALLOCATED is set, it looks like fi_mr_reg will be skipped and that should work.

Reply:
@hzhou is there any advantage of registering a NULL buffer for dynamic window creation for other providers which have MPIDI_OFI_ENABLE_MR_ALLOCATED set to false, since I'm guessing it isn't doing anything internally? Why not just register only when base is non-NULL?

Reply:
@nitbhat I believe the NULL registration during dynamic window creation is an optimization that allows MPI_WIN_ATTACH calls to skip additional memory registrations. Essentially it grants access to the the whole virtual memory space.

Reply:
That makes sense, thanks @raffenet and @hzhou. 

I've added a capability set variable for determine if the provider supports registering NULL and use that to skip registration with NULL addresses for CXI. This PR is ready for review. 

I'm also in conversation with the CXI team to find out why FI_MR_ALLOCATED is not set. If that changes in the future, most of this PR can be reverted.